### PR TITLE
[DEV-3820] Fixes SavePrayer update of Saved List

### PR DESCRIPTION
**Pull request by cdanwards**
### What does this PR do, or why is it needed?

Currently when you save a prayer, it is not optimistically added to the savedPrayerList. This PR adds the needed `update` to the `savePrayer` mutation.

### How do I test this PR?

Go to a prayer menu, save a prayer, then make sure it's added to the savedPrayerList in that session.

---

> I am affirming this is my _best_ work ([Ecclesiastes 9:10](https://www.bible.com/bible/97/ECC.9.10.MSG))

## TODO

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [DEV-XXX](#url)
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android if applicable
- [ ] Set two relevant reviewers

## REVIEW

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

---

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 1]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
 import { Mutation } from 'react-apollo';
 
 import savePrayerRequest from '../data/mutations/savePrayerRequest';
+import savedPrayerList from '../data/queries/getSavedPrayers';
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
import GET_SAVED_PRAYERS from '../data/queries/getSavedPrayers';
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 2]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
...
 import PrayerActionMenuCard from './PrayerActionMenuCard';
 
 const PrayerActionMenuCardConnected = memo(
-  ({ onAdvancePrayer, navigation, prayerId, ...props }) => {
+  ({ onAdvancePrayer, navigation, prayerId, prayerRequest, ...props }) => {
     const exitPrayer = () => navigation.navigate('Prayer');
     return (
-      <Mutation mutation={savePrayerRequest}>
+      <Mutation
+        mutation={savePrayerRequest}
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
        mutation={SAVE_PRAYER}
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 3]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
...
+        update={async (cache) => {
+          const currentSavedPrayers = cache.readQuery({
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
          const { savedPrayers } = cache.readQuery({
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 4]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
...
+            query: savedPrayerList,
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
            query: GET_SAVED_PRAYERS,
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 5]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
...
+          });
+          const newPrayersList = currentSavedPrayers.savedPrayers;
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
          const newPrayersList = savedPrayers;
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 6]
```diff
@@ -3,13 +3,27 @@ import PropTypes from 'prop-types';
...
+          newPrayersList.push(prayerRequest.prayer);
+          await cache.writeQuery({
+            query: savedPrayerList,
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
            query: GET_SAVED_PRAYERS,
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 7]
```diff
@@ -32,6 +46,7 @@ const PrayerActionMenuCardConnected = memo(
 PrayerActionMenuCardConnected.propTypes = {
   onAdvancePrayer: PropTypes.func,
   prayerId: PropTypes.string,
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
```

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js [hunk 8]
```diff
@@ -99,6 +99,7 @@ class PrayerCard extends PureComponent {
       prayerId,
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js]:
```suggestion
```

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js [hunk 9]
```diff
@@ -99,6 +99,7 @@ class PrayerCard extends PureComponent {
...
       navigation,
       anonymous,
+      prayerRequest,
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js:101]:
```suggestion
```

**Review by IsaacHardy — CHANGES_REQUESTED** [review]
placeholder

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js [hunk 10]
```diff
@@ -119,8 +120,9 @@ class PrayerCard extends PureComponent {
         <PrayerActionMenuCardConnected
           expandedHeight={expandedHeight}
           navigation={this.props.navigation}
-          prayerId={prayerId}
           advancePrayer={advancePrayer}
+          prayerId={prayerId}
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js]:
```suggestion
```

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js [hunk 11]
```diff
@@ -119,8 +120,9 @@ class PrayerCard extends PureComponent {
...
+          prayerRequest={prayerRequest}
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js:123]:
```suggestion
         { ...this.props }
```

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js [hunk 12]
```diff
@@ -204,6 +206,7 @@ PrayerCard.propTypes = {
   advancePrayer: PropTypes.func,
   incrementPrayer: PropTypes.func,
   prayerId: PropTypes.string,
+  prayerRequest: PropTypes.shape({}),
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js:206]:
```suggestion
```

File: packages/newspringchurchapp/src/prayer/PrayerList/index.js [hunk 13]
```diff
@@ -191,6 +191,7 @@ class PrayerList extends PureComponent {
                     </DividerView>
                   )}
                   data={get(data, prayers, []).map((prayer) => ({
+                    prayerRequest: { prayer },
```
Comment (redreceipt) [inline packages/newspringchurchapp/src/prayer/PrayerList/index.js]:
```suggestion
```

**Review by redreceipt — CHANGES_REQUESTED** [review]

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 1 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: Holding off on this until Vincent speaks his peace.

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 2 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: Holding off on this until Vincent says his peace

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 3 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: 👍

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 4 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: Holding off on this until Vincent says his peace

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 5 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: 👍

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 6 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: Holding off on this until Vincent says his peace

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js (hunk 7 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]: 👍

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js (hunk 8 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js]: 👍

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js (hunk 10 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js]: 👍

File: packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js (hunk 12 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/prayer/PrayerCard/PrayerCard.js:206]: This is still needed because of the `props.prayerRequest.id`

**Review by redreceipt — DISMISSED** [review]

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 14]
```diff
@@ -1,22 +1,42 @@
...
+import { unionBy } from 'lodash';
 import { Mutation } from 'react-apollo';
 
 import SAVE_PRAYER from '../data/mutations/savePrayer';
+import GET_SAVED_PRAYERS from '../data/queries/getSavedPrayers';
 import PrayerActionMenuCard from './PrayerActionMenuCard';
 
 const PrayerActionMenuCardConnected = memo(
-  ({ onAdvancePrayer, navigation, prayerId, ...props }) => {
+  ({ onAdvancePrayer, navigation, ...props }) => {
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
  ({ onAdvancePrayer, navigation, prayerRequest, ...props }) => {
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 15]
```diff
@@ -1,22 +1,42 @@
...
+        mutation={SAVE_PRAYER}
+        update={async (cache) => {
+          const { savedPrayers } = cache.readQuery({
+            query: GET_SAVED_PRAYERS,
+          });
+          // This methods compares the incoming prayer by id to the items in
+          // the array and if it's not in the array, it adds it.
+          const newPrayersList = unionBy(
+            savedPrayers,
+            [props.prayerRequest],
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
            [prayerRequest],
```

File: packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js [hunk 16]
```diff
@@ -1,22 +1,42 @@
...
+        }}
+      >
         {(savePrayer) => (
           <PrayerActionMenuCard
             exitPrayer={exitPrayer}
             savePrayer={async () => {
               await savePrayer({
                 variables: {
-                  nodeId: prayerId,
+                  nodeId: props.prayerRequest.id,
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/PrayerActionMenuCard/PrayerActionMenuCardConnected.js]:
```suggestion
                  nodeId: prayerRequest.id,
```

**Review by IsaacHardy — DISMISSED** [review]

**Review by IsaacHardy — APPROVED** [review]
